tests/integration/deploy_jenkins.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Deploy jenkins to lxd using juju."""
+
+import dataclasses
+import logging
+import subprocess
+import sys
+import time
+
+import yaml
+from selenium.webdriver import Firefox, FirefoxOptions
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+
+logging.basicConfig(
+    stream=sys.stdout, format="%(levelname)s %(asctime)s - %(message)s", level=logging.INFO
+)
+LOGGER = logging.getLogger()
+
+
+@dataclasses.dataclass
+class JenkinsDeployment:
+    """Key information about a jenkins deployment."""
+
+    controller_name: str
+    model_name: str
+    unit_number: int
+    public_address: str
+    username: str
+    password: str
+
+    @property
+    def hostname(self) -> str:
+        """Calculate the hostname."""
+        if ":" in self.public_address:
+            return f"[{self.public_address}]"
+        return self.public_address
+
+
+def jenkins_active() -> bool:
+    """Check whether the jenkins application is active."""
+    result = subprocess.check_output(["juju", "status"])
+    LOGGER.info("juju status \n%s", result.decode("utf-8"))
+    result = subprocess.check_output(["juju", "status", "--format", "yaml"])
```

Comment:
You should be able to do `app.status == ActiveStatus.name` instead

Reply:
Currently, `deploy_jenkins.py` is not using during test execution and instead used by the CI to get Jenkins deployed, which means `pytest-oprtator` functionality isn't available. The primary reason it was designed like that was because spinning up jenkins servers locally is expensive. This meant that locally I had jenkins running separately and passed arguments for the controller, model and unit number into the tests and in CI, Jenkins gets setup and again information about the controller, model and unit number are passed in as parameters to the test.

The alternative is to change the design to make `setup_jenkins.py` a fixture that is only run if the parameters for jenkins are not passed in, as an example.

I can see the advantages of turning this into a fixture, it would avoid some complexities in CI. It would probably take a bit more effort to get that to be well designed, maybe the equivalent of 3 points. It would be quite interesting to get it to work.

Reply:
Understood. Thank you for clarifying.
Anyway, it'd certainly fall out of the scope of the PR.

Reply:
I'll add a new ticket to the backlog